TECH-5214: use escalate in exception_handling

ColinDKelley opened the pull request:
## [0.2.0] - Unreleased
### Added
- Added support for `on_escalate(log_first: false)`. The `escalate` gem will log first before
  escalating if either of these is true: there are no `on_escalate` blocks registered, or
  there are some and at least one of them passed `log_first: false`.

jebentier commented on lib/escalate.rb:
```diff
@@ -81,8 +83,20 @@ def default_escalate_logger
+    # @param [boolean] log_first: true
+    #   whether escalate should log first before escalating, or leave the logging to the escalate block
+    def on_escalate(log_first: true, &block)
+      if log_first
```
Does this need to be split into two separate groups? Can we simplify this by having a private class variable defined `disable_logging` which starts as `false` and then when `log_first` comes in as `false` it flips to `true`? I think overall this would reduce the complexity of holding two sets of callbacks.

```
def on_escalate(log_first: true, &block)
  disable_logging = self.disable_logging || !log_first
  on_escalate_blocks.add(block)
end
```

ColinDKelley replied:
I didn't want to be stateful about the order of callback declaration.

ColinDKelley replied:
What I really wanted was to store an attribute on the callbacks. Turns out that Ruby actually allows this. What do you think?
```
block.instance_variable_set(:@log_first, log_first)
```
Then later I could check those:
```
on_escalate_blocks.any? { |block| block.instance_variable_get(:@log_first) }
```

ColinDKelley replied:
@jebentier I took a shot at it here: https://github.com/Invoca/escalate/pull/2/commits/2e946ae65f5802d588e5d022bcd811a0ade42fd9

Seems much cleaner to me. It's slightly thin ice to put instance variables on blocks since Ruby doesn't document that as a feature. But these unit tests will notice if they ever retract that. Seem ok?

jebentier replied:
> I didn't want to be stateful about the order of callback declaration.

I get what you're talking about now with this comment now after seeing this new implementation.

> Seems much cleaner to me. It's slightly thin ice to put instance variables on blocks since Ruby doesn't document that as a feature. But these unit tests will notice if they ever retract that. Seem ok?

Definitely agree that this seems much cleaner. I'm good with this thin contract. We can always think about different approaches later as well if we are concerned with the thin nature of this practice.

ColinDKelley replied:
Cool. I feel pretty good with the "thin ice" here since it's completely encapsulated inside this gem, with automated tests to prove it works.
Hmm, I wonder if it'll work in JRuby?!

jebentier replied:
🤷

dcaddell commented on lib/escalate.rb:
```diff
@@ -9,37 +9,41 @@
+  LOG_FIRST_INSTANCE_VARIABLE = :@_escalate_log_first
+
   class << self
     # Logs and escalated an exception
```
This documentation seems like it should be updated to be a bit more descriptive since the contract is more complicated now. I think it's especially worth calling out that we **won't do any logging** if an escalation callback is registered without `log_first: true`.

ColinDKelley replied:
Good catch! I'll update the docs tomorrow before merging.

ColinDKelley commented:
@dcaddell and @jebentier I updated the README and renamed the callbacks to match the documentation. Also I added a test for the non-contextual logger, and changed the contextual one not to include the context hash in the message since it'll appear in the JSON. Can you review those changes?

ColinDKelley commented:
@jebentier Can you please review latest changes here per Slack discussion?
